### PR TITLE
event - add kafka support (publish/subscribe)

### PR DESCRIPTION
**Pull request by stfung77**
add support for the anchor platform to publish events to Kafka
add sending events for quote_created, transaction_created, transaction_payment_received

add support for reference server event consumer to read events from Kafka
add support for reference server event consumer to process transaction_payment_received events and update the transaction status on the anchor platform side via PATCH

add example docker-compose file for running single instance kafka

addresses:
https://github.com/stellar/java-stellar-anchor-sdk/issues/157
https://github.com/stellar/java-stellar-anchor-sdk/issues/155

### CODE REVIEW

File: core/src/main/java/org/stellar/anchor/event/models/QuoteEvent.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+  @SerializedName("created_at")
+  LocalDateTime createdAt;
+
+  @SerializedName("client_domain")
+  String clientDomain;
+
+  public QuoteEvent() {}
+
+  public QuoteEvent(
```
Comment (lijamie98) [inline core/src/main/java/org/stellar/anchor/event/models/QuoteEvent.java:37]: Consider `@AllArgsConstructor`

File: core/src/main/java/org/stellar/anchor/event/models/TransactionEvent.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
+package org.stellar.anchor.event.models;
+
+import com.google.gson.annotations.SerializedName;
+import lombok.Data;
+
+import java.time.LocalDateTime;
+
+@Data
+public class TransactionEvent implements AnchorEvent {
```
Comment (lijamie98) [inline core/src/main/java/org/stellar/anchor/event/models/TransactionEvent.java:9]: `@AllArgsConstructor` can help.

File: core/src/main/java/org/stellar/anchor/sep38/Sep38Service.java [hunk 3]
```diff
@@ -368,7 +373,21 @@ public Sep38QuoteResponse postQuote(JwtToken token, Sep38PostQuoteRequest reques
             .build();
     this.sep38QuoteStore.save(newQuote);
 
-    // TODO: create an event for `quote_created` using the event API
+    // TODO: use constant instead of "quote_created"
+    QuoteEvent event =
```
Comment (lijamie98) [inline core/src/main/java/org/stellar/anchor/sep38/Sep38Service.java:377]: `@AllArgsConstructor` or `@Builder` may help.

Comment (lijamie98) [reply core/src/main/java/org/stellar/anchor/sep38/Sep38Service.java:377]: https://github.com/stellar/java-stellar-anchor-sdk/blob/main/core/src/main/java/org/stellar/anchor/dto/sep31/Sep31GetTransactionResponse.java#L12

File: anchor-reference-server/src/main/java/org/stellar/anchor/reference/AnchorEventServer.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+import java.util.Properties;
+
+@Component
+public class AnchorEventServer implements DisposableBean, Runnable {
+    private Thread thread;
+    private Consumer consumer;
+    private volatile boolean shutdown = false;
+
+    AnchorEventServer(){
+        this.thread = new Thread(this);
```
Comment (lijamie98) [inline anchor-reference-server/src/main/java/org/stellar/anchor/reference/AnchorEventServer.java:29]: The Java ExecutorService provides thread pool and easier to manage than the Thread. This may help the stability of the server.

https://www.baeldung.com/java-executor-service-tutorial

File: anchor-reference-server/src/main/java/org/stellar/anchor/reference/AnchorEventServerConfig.java [hunk 5]
```diff
@@ -0,0 +1,4 @@
+package org.stellar.anchor.reference;
+
+public class AnchorEventServerConfig {
+}
```
Comment (marcelosalloum) [inline anchor-reference-server/src/main/java/org/stellar/anchor/reference/AnchorEventServerConfig.java:4]: Is this file needed?

File: anchor-reference-server/src/main/java/org/stellar/anchor/reference/AnchorEventServer.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+import java.io.UnsupportedEncodingException;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.Properties;
+
+@Component
+public class AnchorEventServer implements DisposableBean, Runnable {
```
Comment (marcelosalloum) [inline anchor-reference-server/src/main/java/org/stellar/anchor/reference/AnchorEventServer.java:23]: Maybe we could go with a more descriptive name for AnchorEventServer, maybe `AnchorEventConsumerService`?

File: anchor-reference-server/src/main/java/org/stellar/anchor/reference/AnchorEventServer.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+    AnchorEventServer(){
+        this.thread = new Thread(this);
+        this.thread.start();
+    }
+
+    @Override
+    public void run(){
+
+        System.out.println("queue consumer server started ");
```
Comment (marcelosalloum) [inline anchor-reference-server/src/main/java/org/stellar/anchor/reference/AnchorEventServer.java:36]: I think you should have access to Log.info from this file. It would be better to log things using Log.info when we're in production.

File: anchor-reference-server/src/main/java/org/stellar/anchor/reference/AnchorEventServer.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+                                    default:
+                                        System.out.printf("error: anchor_platform_event - invalid " +
+                                                "message type '%s'%n", eventType);
+                                }
+                                break;
+                            }
+                        }
+                    });
+            } catch (Exception ex) {
+                System.out.println(ex.toString());
```
Comment (marcelosalloum) [inline anchor-reference-server/src/main/java/org/stellar/anchor/reference/AnchorEventServer.java:91]: Log.error

File: core/src/main/java/org/stellar/anchor/config/EventConfig.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package org.stellar.anchor.config;
+
+
+import java.util.Map;
+
+public interface EventConfig {
+  Map<String, String> getQueues();
+
+  String getKafkaBootstrapServer();
```
Comment (marcelosalloum) [inline core/src/main/java/org/stellar/anchor/config/EventConfig.java:9]: These names are confusing to me...

getKafkaBootstrapServer returns a string instead of an error, so shouldn't it be something like `getKafkaBootstrapServerId()` or `getKafkaBootstrapServerName()`?

Also, getQueues() doesn't return a queue nor a queue map. What exactly is it about?

File: core/src/main/java/org/stellar/anchor/event/models/TransactionEvent.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+  StellarId creator;
+
+  public enum Status {
+    PENDING_SENDER("pending_sender"),
+    PENDING_STELLAR("pending_stellar"),
+    PENDING_CUSTOMER_INFO_UPDATE("pending_customer_info_update"),
+    PENDING_RECEIVER("pending_receiver"),
+    PENDING_EXTERNAL("pending_external"),
+    COMPLETED("completed"),
+    ERROR("error");
```
Comment (marcelosalloum) [inline core/src/main/java/org/stellar/anchor/event/models/TransactionEvent.java:88]: You're missing `pending_transaction_info_update`.

File: platform/src/main/java/org/stellar/anchor/platform/service/PaymentOperationToEventListener.java [hunk 11]
```diff
@@ -66,30 +73,55 @@ public void onSent(PaymentOperationResponse payment) {
...
+      .id(txn.getId())
+      .status(TransactionEvent.Status.valueOf(txn.getStatus()))
+      .sep(org.stellar.anchor.event.models.TransactionEvent.Sep.SEP_31)
+      .kind(org.stellar.anchor.event.models.TransactionEvent.Kind.RECEIVE)
+      .amountIn(org.stellar.anchor.event.models.Amount.builder()
+              .amount(txn.getAmountIn())
+              .asset(txn.getAmountInAsset())
+              .build()
+      )
+      .amountOut(org.stellar.anchor.event.models.Amount.builder()
```
Comment (marcelosalloum) [inline platform/src/main/java/org/stellar/anchor/platform/service/PaymentOperationToEventListener.java:93]: Do you mind removing the `org.stellar.anchor.event.models..*` full paths and importing the classes on the top of the file instead?

**Review by marcelosalloum — APPROVED** [review]
LGTM! I added a few suggestions below.

File: core/src/main/java/org/stellar/anchor/config/EventConfig.java (hunk 9 above)
Comment (stfung77) [reply core/src/main/java/org/stellar/anchor/config/EventConfig.java:9]: getQueues() returns a map of event type to queue/topic name

File: anchor-reference-server/src/main/java/org/stellar/anchor/reference/AnchorEventServer.java (hunk 4 above)
Comment (stfung77) [reply anchor-reference-server/src/main/java/org/stellar/anchor/reference/AnchorEventServer.java]: will do this in a separate PR

File: build.gradle.kts [hunk 12]
```diff
@@ -13,6 +13,7 @@ subprojects {
   repositories {
     mavenCentral()
     maven { url = uri("https://jitpack.io") }
+    maven { url = uri("https://packages.confluent.io/maven") }
```
Comment (lijamie98) [inline build.gradle.kts:16]: Is this for Kafka dependencies?

File: build.gradle.kts [hunk 13]
```diff
@@ -49,6 +50,13 @@ subprojects {
   dependencies {
     // This is to fix the missing implementation in JSR305 that causes "unknown enum constant When.MAYBE" warning.
     implementation("com.google.code.findbugs:jsr305:3.0.2")
+    implementation("org.apache.kafka:kafka-clients:3.1.0")
+    implementation("org.apache.kafka:connect:3.1.0")
+    // https://mvnrepository.com/artifact/io.confluent/kafka-json-schema-serializer
```
Comment (lijamie98) [inline build.gradle.kts:55]: remove?

File: core/src/main/java/org/stellar/anchor/event/models/Amount.java [hunk 14]
```diff
@@ -0,0 +1,15 @@
+package org.stellar.anchor.event.models;
```
Comment (lijamie98) [inline core/src/main/java/org/stellar/anchor/event/models/Amount.java:1]: Can we use https://github.com/stellar/java-stellar-anchor-sdk/blob/main/platform-apis/src/main/java/org/stellar/platform/apis/shared/Amount.java instead?

Comment (lijamie98) [inline core/src/main/java/org/stellar/anchor/event/models/Amount.java:1]: These classes should belong to the `platform-api`. We may merge this and move the classes in the next PR.

File: core/src/main/java/org/stellar/anchor/event/models/TransactionEvent.java [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+
+  Type type;
+
+  public String getType() {
+    return this.type.type;
+  }
+
+  String id;
+
+  Sep sep;
```
Comment (lijamie98) [inline core/src/main/java/org/stellar/anchor/event/models/TransactionEvent.java:25]: Can we use integer instead of the `Sep` class?

File: platform/src/main/java/org/stellar/anchor/platform/service/PaymentOperationToEventListener.java [hunk 16]
```diff
@@ -93,33 +102,56 @@ public void onSent(ObservedPayment payment) {
...
-        .build();
+    // TODO move event models to /shared
+    TransactionEvent event =
+        TransactionEvent.builder()
+            .eventId(UUID.randomUUID().toString())
+            .type(TransactionEvent.Type.TRANSACTION_PAYMENT_RECEIVED)
+            .id(txn.getId())
+            .status(TransactionEvent.Status.PENDING_RECEIVER)
+            .sep(org.stellar.anchor.event.models.TransactionEvent.Sep.SEP_31)
+            .kind(org.stellar.anchor.event.models.TransactionEvent.Kind.RECEIVE)
```
Comment (marcelosalloum) [inline platform/src/main/java/org/stellar/anchor/platform/service/PaymentOperationToEventListener.java:118]: So cool that you enriched the object.

File: core/src/main/java/org/stellar/anchor/event/models/TransactionEvent.java [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+
+    @JsonValue
+    public Integer getSep() {
+      return sep;
+    }
+  }
+
+  public enum Type {
+    TRANSACTION_CREATED("transaction_created"),
+    TRANSACTION_PAYMENT_RECEIVED("transaction_payment_received"),
```
Comment (marcelosalloum) [inline core/src/main/java/org/stellar/anchor/event/models/TransactionEvent.java:115]: @lijamie98 has just updated this event name to `transaction_status_changed` in https://github.com/stellar/stellar-anchor-platform/pull/31:
```suggestion
    TRANSACTION_STATUS_CHANGED("transaction_status_changed"),
```

**Review by lijamie98 — COMMENTED** [review]
LGTM. There are some minor comments.

**Review by lijamie98 — APPROVED** [review]

File: anchor-reference-server/src/main/java/org/stellar/anchor/reference/AnchorEventConsumerService.java [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+                                                    transactionEvent.getAmountFee().getAmount(),
+                                                    transactionEvent.getAmountFee().getAsset()))
+                                            .amountOut(
+                                                new Amount(
+                                                    transactionEvent.getAmountOut().getAmount(),
+                                                    transactionEvent.getAmountOut().getAsset()))
+                                            .build()))
+                                .build();
+                        PlatformApiClient platformClient =
+                            new PlatformApiClient("http://localhost:8080");
```
Comment (marcelosalloum) [inline anchor-reference-server/src/main/java/org/stellar/anchor/reference/AnchorEventConsumerService.java:104]: You're mocking the transaction completion, right? Can you add a comment making it clear?

**Review by marcelosalloum — APPROVED** [review]
LGTM! Feel free to merge it after addressing the comments below.

File: core/src/main/java/org/stellar/anchor/event/models/TransactionEvent.java (hunk 17 above)
Comment (stfung77) [reply core/src/main/java/org/stellar/anchor/event/models/TransactionEvent.java:115]: lets address that in a separate PR since it requires a schema change

File: core/src/main/java/org/stellar/anchor/event/models/TransactionEvent.java (hunk 15 above)
Comment (stfung77) [reply core/src/main/java/org/stellar/anchor/event/models/TransactionEvent.java:25]: i'll address this in a separate PR where i'll also update the transaction_status_changed schema

File: core/src/main/java/org/stellar/anchor/event/models/Amount.java (hunk 14 above)
Comment (stfung77) [reply core/src/main/java/org/stellar/anchor/event/models/Amount.java:1]: will refactor in a separate PR

Comment (stfung77) [reply core/src/main/java/org/stellar/anchor/event/models/Amount.java:1]: will refactor in a separate PR

File: build.gradle.kts (hunk 12 above)
Comment (stfung77) [reply build.gradle.kts:16]: yes
